src/variational/advi.jl

```diff
@@ -0,0 +1,213 @@
+
+        # initials
+        μ_i = μ[r]
+        ω_i = ω[r]
+
+        # # sample from VI posterior
+        η = randn(rng, length(μ_i))
+        ζ = center_diag_gaussian_inv(η, μ_i, exp.(ω_i))
+        θ = invlink(prior, ζ)
+
+        x[r] = θ
+    end
+
+    return x
+end
+
+"""
+    ADVI(samplers_per_step = 10, max_iters = 5000)
+
+Automatic Differentiation Variational Inference (ADVI) for a given model.
+"""
+struct ADVI <: VariationalInference
+    samples_per_step # number of samples used to estimate the ELBO in each optimization step
+    max_iters        # maximum number of gradient steps used in optimization
+end
+
+ADVI() = ADVI(10, 5000)
+
+alg_str(::ADVI) = "ADVI"
+
+vi(model::Model, alg::ADVI; optimizer = ADAGrad()) = begin
+    # setup
+    var_info = VarInfo()
+    model(var_info, SampleFromUniform())
+    num_params = size(var_info.vals, 1)
+
+    dists = var_info.dists
+    ranges = var_info.ranges
+    
+    q = MeanField(zeros(num_params), zeros(num_params), dists, ranges)
+
+    # construct objective
+    elbo = ELBO()
+
+    Turing.DEBUG && @debug "Optimizing ADVI..."
+    μ, ω = optimize(elbo, alg, q, model; optimizer = optimizer)
+
+    # TODO: make mutable instead?
+    MeanField(μ, ω, dists, ranges) 
+end
+
+# TODO: implement optimize like this?
+# (advi::ADVI)(elbo::EBLO, q::MeanField, model::Model) = begin
+# end
+
+function optimize(elbo::ELBO, alg::ADVI, q::MeanField, model::Model; optimizer = ADAGrad())
+    alg_name = alg_str(alg)
+    samples_per_step = alg.samples_per_step
+    max_iters = alg.max_iters
+
+    # number of previous gradients to use to compute `s` in adaGrad
+    stepsize_num_prev = 10
+    
+    # setup
+    var_info = Turing.VarInfo()
+    model(var_info, Turing.SampleFromUniform())
+    num_params = size(var_info.vals, 1)
+
+    function f(x)
+        # extract the mean-field Gaussian params
+        μ, ω = x[1:num_params], x[num_params + 1: end]
+        
+        - elbo(q, model, μ, ω, samples_per_step)
+    end
+
+    # buffer
+    x = zeros(2 * num_params)
+
+    # HACK: re-use previous gradient `acc` if equal in value
+    # Can cause issues if two entries have idenitical values
+    vs = [v for v ∈ keys(optimizer.acc)]
+    idx = findfirst(w -> vcat(q.μ, q.ω) == w, vs)
+    if idx != nothing
+        @info "[$alg_name] Re-using previous optimizer accumulator"
+        x = vs[idx]
+    end
+    
+    diff_result = DiffResults.GradientResult(x)
+
+    # TODO: in (Blei et al, 2015) TRUNCATED ADAGrad is suggested; this is not available in Flux.Optimise
+    # Maybe consider contributed a truncated ADAGrad to Flux.Optimise
+
+    i = 0
+    prog = PROGRESS[] ? ProgressMeter.Progress(max_iters, 1, "[$alg_name] Optimizing...", 0) : 0
+
+    time_elapsed = @elapsed while (i < max_iters) # & converged # <= add criterion? A running mean maybe?
+        # TODO: separate into a `grad(...)` call; need to manually provide `diff_result` buffers
+        ForwardDiff.gradient!(diff_result, f, x)
```

Comment:
Turing has a `grad` function, maybe this could be adjusted so that you can use it here instead?

Reply:
Interested! Could you elobarate? I can only find the gradients of `logp`. 

Also, the comment I've written there is referring to the fact that I should follow the interface suggested in this PR (see above). Nonetheless, in _that_ function I'll need to compute the gradient.

Reply:
Sure, I was referring to looking into the grad function of Turing and to adjust / make a new one similar to the grad function we have. This would help to make your code less dependent on the AD framework.

Reply:
Ah, follow :+1:  Somehow thought "this could be adjusted" referred to adjusting the above implementation to _use_ the already existing function; but you're suggesting adjusting existing gradient functions to _this_. Makes much more sense!:) 

I'll definitively try and do that; thanks!

Reply:
So I've managed to support both `ForwardDiff` and `Tracker`, but the `Tracker` implementation runs _very_ slowly.

For the `ADVI` case I believe the optimization can be sped up quite a bit by simply removing the `grad!` call and performing the gradient computation directly in `optimize!`. This way we can define the `θ_tracked` outside of the loop, and use this for inplace gradient computation using `Tracker`. For `ForwardDiff`, we can define the objective function outside of the loop capturing the parameters by closing over them. This circumvents redefining the objective function upon each step.

Anyways, the current implementation is slow but works as expected.